alibi/explainers/base.py

```diff
@@ -0,0 +1,127 @@
+import abc
+import copy
+import json
+from collections import ChainMap
+from typing import Any
+import logging
+
+import attr
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+# default data and metadata
+DEFAULT_META = {
+    "name": None,
+    "type": [],
+    "explanations": [],
+    "params": {},
+}  # type: dict
+
+DEFAULT_DATA = {}  # type: dict
+
+
+@attr.s
+class Explainer(abc.ABC):
+    """
+    Base class for explainer algorithms
+    """
+
+    meta = attr.ib(default=copy.deepcopy(DEFAULT_META))  # type: dict
+
+    def __attrs_post_init__(self):
+        self.meta["name"] = self.__class__.__name__
+
+    @abc.abstractmethod
+    def explain(self, X: Any) -> "Explanation":
+        pass
+
+
+class FitMixin(abc.ABC):
+    @abc.abstractmethod
+    def fit(self, X: Any) -> "Explainer":
+        pass
+
+
+@attr.s
+class Explanation:
```

Comment:
I agree that this would be a great feature. I'm not sure using `__dict__` is the best approach as it will have lots of unrelated stuff in, but we can fetch the list of attributes in the same way as we do when exposing them dynamically in `__attrs_post_init__` and set it as a property.

Reply:
so presumably that can be the implementation of `attributes`

Reply:
Currently not committing to implementing/maintaining as its easy-ish to get for the user.